leap_c/torch/rl/ppo.py

```diff
@@ -0,0 +1,354 @@
+            self.cfg.actor_mlp
+        )
+        self.pi_optim = torch.optim.Adam(self.pi.parameters(), lr=self.cfg.lr_pi)
+        self.pi_lr_scheduler = torch.optim.lr_scheduler.LinearLR(
+            self.pi_optim,
+            start_factor=1.0,
+            end_factor=0.0,
+            total_iters=self.cfg.train_steps // (self.cfg.num_steps * self.train_env.num_envs)
+        )
+
+        self.clipped_loss = ClippedSurrogateLoss(self.cfg.clipping_epsilon)
+        self.value_loss = ValueSquaredErrorLoss(self.cfg.clip_value_loss, self.cfg.clipping_epsilon)
+
+        self.buffer = ReplayBuffer(self.cfg.num_steps, device)
+
+    def train_loop(self) -> Iterator[int]:
+        obs, _ = self.train_env.reset(seed=self.cfg.seed, options={"mode": "train"})
+
+        while True:
+            # region Rollout Collection
+            obs_collate = torch.tensor(obs, device=self.device)
+
+            with torch.no_grad():
+                action, log_prob, _ = self.pi(obs_collate)
+                value = self.q(obs_collate)
+
+            value = value.cpu().numpy()
+            action = action.cpu().numpy()
+            log_prob = log_prob.cpu().numpy()
+
+            self.report_stats("train_trajectory", {"action": action}, verbose=True, with_smoothing=False)
+
+            obs_prime, reward, is_terminated, is_truncated, info = self.train_env.step(action)
+
+            self.buffer.put((
+                obs,
+                action,
+                log_prob,
+                reward,
+                obs_prime,
+                is_terminated,
+                np.logical_or(is_terminated, is_truncated),
+                value
+            ))
+
+            if "episode" in info:
+                idx = info["_episode"].argmax()
+                self.report_stats("train", {
+                    "episodic_return": float(info["episode"]["r"][idx]),
+                    "episodic_length": int(info["episode"]["l"][idx]),
+                }, with_smoothing=False)
+            # endregion
+
+            obs = obs_prime
+
+            if (self.state.step + self.train_env.num_envs) % (self.cfg.num_steps * self.train_env.num_envs) == 0:
+                # region Generalized Advantage Estimation (GAE)
+                advantages = torch.zeros((self.cfg.num_steps, self.train_env.num_envs), device=self.device)
+                returns = torch.zeros((self.cfg.num_steps, self.train_env.num_envs), device=self.device)
+                with torch.no_grad():
+                    for t in reversed(range(self.cfg.num_steps)):
+                        _, _, _, reward, obs_prime, termination, done, value = self.buffer[t]
+
+                        reward = reward.squeeze(0)
+                        obs_prime = obs_prime.squeeze(0)
+                        termination = termination.squeeze(0)
+                        done = done.squeeze(0)
+                        value = value.squeeze(0)
+
+                        value_prime = self.q(obs_prime)
```

Comment:
the critic isn't updated between data collection and GAE calculation, so no need to use `self.q`, can use `values[t + 1]` directly to make the algorithm faster.